File: src/day19b.py
# Copyright (c) 2021 Mikołaj Kuranowski
# SPDX-License-Identifier: WTFPL
from typing import Callable, Iterable
from fileinput import FileInput

import core

RuleMatcher = Callable[[str, "RuleMatchers"], set[str]]
RuleMatchers = dict[str, "RuleMatcher"]


def parse_conjunction_rule(rule: str, name: str) -> RuleMatcher:
    rule_names = rule.split()

    def matcher(text: str, matchers: RuleMatchers) -> set[str]:
        inputs: set[str] = {text}
        outputs: set[str] = set()

        for rule_name in rule_names:

            for input in inputs:
                outputs.update(matchers[rule_name](input, matchers))

            # No more matches - no need to process anymore
            if not outputs:
                return outputs

            inputs = outputs
            outputs = set()

        return inputs

    matcher.__name__ = f"matcher-{name}"
    return matcher


def parse_disjunction_rule(rule: str, name: str) -> RuleMatcher:
    alternatives = [
        parse_conjunction_rule(i, f"{name}-alt{idx}")
        for idx, i in enumerate(rule.split(" | "))
    ]

    def matcher(text: str, matchers: RuleMatchers) -> set[str]:
        all_outputs: set[str] = set()

        for alternative in alternatives:
            all_outputs.update(alternative(text, matchers))

        return all_outputs

    matcher.__name__ = f"matcher-{name}"
    return matcher


def parse_char_rule(rule: str, name: str) -> RuleMatcher:
    assert rule[0] == rule[2] == '"'
    char = rule[1]

    def matcher(text: str, matchers: RuleMatchers) -> set[str]:
        return {text[1:]} if text[:1] == char else set()

    matcher.__name__ = f"matcher-{name}"
    return matcher


def parse_rule(rule: str, name: str) -> RuleMatcher:
    if '"' in rule:
        return parse_char_rule(rule, name)
    elif "|" in rule:
        return parse_disjunction_rule(rule, name)
    else:
        return parse_conjunction_rule(rule, name)


def parse_rules(lines: Iterable[str]) -> RuleMatchers:
    matchers: RuleMatchers = {}
    for line in lines:
        name, _, rule_text = line.partition(": ")
        matchers[name] = parse_rule(rule_text, name)
    return matchers


def matches_zero_fully(text: str, matchers: RuleMatchers) -> bool:
    result = matchers["0"](text, matchers)
    return "" in result


if __name__ == "__main__":
    input: "FileInput[str]" = FileInput()
    rules, lines = core.split_on((i.rstrip() for i in input), core.empty_str)
    matchers = parse_rules(rules)
    result = sum(1 for line in lines if matches_zero_fully(line, matchers))
    print(result)

File: src/day19b.clj
; Copyright (c) 2022 Mikołaj Kuranowski
; SPDX-License-Identifier: WTFPL
(ns day19b
  (:require [core]
            [clojure.string :refer [split includes?]]
            [clojure.set :refer [union]]))

; A "matcher" is a predicate with the following signature:
; (String, Map<String, matcher>) -> Set<String>
;
; A matcher must return a set of all possible matches, to accomodate recursive/loopy rules
;
; On no match, an empty set must be returned.
; On a full match, the returned set must contain "".

(defn parse-conjunction-rule
  [rule]
  (let [rule-numbers (split rule #" ")]
    (fn [text matchers]
      ; Loop over every matcher in the conjunction, starting with the input set only containing the input text
      (loop [matcher (matchers (first rule-numbers))
             rest-rules (rest rule-numbers)
             inputs #{text}]
        (if (some? matcher)
          ; Move to the next matcher, updating the input set
          ; with every possible output of the matcher on every possible input.
          ; Note that on no match, an empty set will be passed through.
          (recur (matchers (first rest-rules))
                 (rest rest-rules)
                 (apply union (map #(matcher % matchers) inputs)))
          ; End of the loop
          inputs)))))

(defn parse-disjunction-rule
  [rule]
  (let [alternatives (map parse-conjunction-rule (split rule #" \| "))]
    ; The result of an disjunction rule is a union of the result sets of every alternative
    (fn [text matchers] (apply union (map #(% text matchers) alternatives)))))

(defn parse-char-rule
  [rule]
  (let [c (second rule)]
    (fn [text _] (if (= (first text) c) #{(apply str (rest text))} #{}))))


(defn parse-rule
  [rule]
  (cond
    (includes? rule "\"") (parse-char-rule rule)
    (includes? rule "|") (parse-disjunction-rule rule)
    :else (parse-conjunction-rule rule)))

(defn parse-rule-line
  [line]
  (let [[_ name rule-text] (re-matches #"(.*): (.*)" line)]
    [name (parse-rule rule-text)]))

(defn parse-rules [lines] (into {} (map parse-rule-line lines)))

(defn matches-zero-fully
  [text matchers]
  (let [r ((matchers "0") text matchers)]
    (contains? r "")))

(defn -main [filename]
  (let [[rule-lines lines] (core/split-on empty? (core/lines-from-file filename))
        matchers (parse-rules rule-lines)]
    (prn (core/count-if #(matches-zero-fully % matchers) lines))))
